sdks/java/core/src/main/java/org/apache/beam/sdk/io/CompressedSource.java

```diff
@@ -149,9 +152,54 @@ public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channe
         return Channels.newChannel(
             new BZip2CompressorInputStream(Channels.newInputStream(channel)));
       }
+    },
+
+    /**
+     * Reads a byte channel assuming it is compressed with zip.
+     * If the zip file contains multiple entries, files in the zip are concatenated all together.
+     */
+    ZIP {
+      @Override
+      public boolean matches(String fileName) {
+        return fileName.toLowerCase().endsWith(".zip");
+      }
+
+      public ReadableByteChannel createDecompressingChannel(ReadableByteChannel channel)
+        throws IOException {
+        FullZipInputStream zip = new FullZipInputStream(Channels.newInputStream(channel));
+        return Channels.newChannel(zip);
+      }
     };
 
     /**
+     * Extend of {@link ZipInputStream} to automatically read all entries in the zip.
+     */
+    public class FullZipInputStream extends ZipInputStream {
+
+      private ZipEntry currentEntry;
+
+      public FullZipInputStream(InputStream is) throws IOException {
+        super(is);
+        currentEntry = getNextEntry();
+      }
+
+      public int read(byte[] b, int off, int len) throws IOException {
+        int result = super.read(b, off, len);
+        if (result == -1) {
+          currentEntry = getNextEntry();
+          if (currentEntry == null) {
+            return -1;
+          }
+          result = super.read(b, off, len);
+          return result;
```

Comment:
I tested the following cases:
1. zip with two entries (non empty)
2. zip with the first entry empty and the second one with data
3. zip with the first entry with data and the second one empty

It works fine:

```
    String filename = createZipFile(new ArrayList<String>(), null, new String[]{}, new
        String[]{"three", "four"});

    FullZipInputStream zipInputStream = new FullZipInputStream(new FileInputStream(new File
        (filename)));
    BufferedReader reader = new BufferedReader(new InputStreamReader(zipInputStream));
    String line;
    while ((line = reader.readLine()) != null) {
      System.out.println(line);
    }
```